src/lib/hash/shake/shake.h

```diff
@@ -9,71 +10,70 @@
 #define BOTAN_SHAKE_HASH_H_
 
 #include <botan/hash.h>
-#include <botan/secmem.h>
+#include <botan/internal/fmt.h>
+#include <botan/internal/shake_xof.h>
 #include <string>
 
 namespace Botan {
 
-/**
-* SHAKE-128
-*/
-class SHAKE_128 final : public HashFunction {
+namespace detail {
+
+template <typename XofT, typename DerivedT>
+class SHAKE_Adapter : public HashFunction {
    public:
-      /**
-      * @param output_bits the desired output size in bits
-      * must be a multiple of 8
-      */
-      explicit SHAKE_128(size_t output_bits);
+      explicit SHAKE_Adapter(size_t output_bits) : m_output_bits(output_bits) {
+         if(output_bits % 8 != 0) {
+            throw Invalid_Argument(fmt("{}: Invalid output length {}", name(), output_bits));
+         }
+      }
 
-      size_t hash_block_size() const override { return SHAKE_128_BITRATE / 8; }
+      size_t hash_block_size() const override { return m_xof.block_size(); }
 
       size_t output_length() const override { return m_output_bits / 8; }
 
-      std::unique_ptr<HashFunction> new_object() const override;
-      std::unique_ptr<HashFunction> copy_state() const override;
-      std::string name() const override;
-      void clear() override;
+      std::string name() const override { return fmt("{}({})", m_xof.name(), m_output_bits); }
+
+      void clear() override { m_xof.clear(); }
+
+      std::unique_ptr<HashFunction> new_object() const override { return std::make_unique<DerivedT>(m_output_bits); }
+
+      std::unique_ptr<HashFunction> copy_state() const override {
+         return std::make_unique<DerivedT>(dynamic_cast<const DerivedT&>(*this));
+      }
 
    private:
-      void add_data(const uint8_t input[], size_t length) override;
-      void final_result(uint8_t out[]) override;
+      void add_data(const uint8_t input[], size_t length) override { m_xof.update({input, length}); }
 
-      static const size_t SHAKE_128_BITRATE = 1600 - 256;
+      void final_result(uint8_t out[]) override {
+         m_xof.output({out, output_length()});
+         m_xof.clear();
+      }
 
+   private:
       size_t m_output_bits;
-      secure_vector<uint64_t> m_S;
-      size_t m_S_pos;
+      XofT m_xof;
 };
 
+}  // namespace detail
+
 /**
-* SHAKE-256
+* SHAKE-128 with a fixed output length
+*
+* For variable output lengths consider using the XOF interface
 */
-class SHAKE_256 final : public HashFunction {
+class SHAKE_128 final : public detail::SHAKE_Adapter<SHAKE_128_XOF, SHAKE_128> {
```

Comment:
All of this seems completely unnecessary. Why not just hold a `std::unique_ptr<XOF>` of the appropriate type?

Reply:
The rationale is similar to [the one discussed back in May](https://github.com/randombit/botan/pull/3525#discussion_r1186276249). I think, we never really came to a conclusion whether this is just a code smell or might cause actual trouble.

Regardless: I decided not to touch the `HashFunction` implementation of SHAKE in this PR at all. With #3675 introducing `Keccak_Permutation::squeeze()`, there's no significant difference in level of abstraction between `Keccak_Permutation` and `SHAKE_XOF`. Hence, I think it's just fine to avoid the indirection via `SHAKE_XOF` here.